sokutei.h

```diff
@@ -0,0 +1,194 @@
+    string_a[SOKUTEI_MAX_COUNTER_NAME_LENGTH] = '\0';
+    return string_a;
+}
+#define sokutei_string_equals(string_a, string_b) (sokutei_strcmp(string_a, string_b) == 0)
+
+
+int sokutei_number_of_iterations = SOKUTEI_MAX_MEASURED_ITERATIONS;
+
+/**
+ * sokutei_counter_definitions is an array consist of the counter name followed by string separator (\0), and the type of the counter
+ */
+char sokutei_counter_definitions[SOKUTEI_MAX_COUNTER_COUNT][SOKUTEI_MAX_COUNTER_NAME_LENGTH + 2] = {'\0'};
+
+
+char sokutei_counters[SOKUTEI_MAX_COUNTER_COUNT * MAX_SIZE_OF_TYPES * SOKUTEI_MAX_MEASURED_ITERATIONS] = {0};
+int  sokutei_number_of_counters = 0;
+int number_of_iterations = SOKUTEI_MAX_MEASURED_ITERATIONS;
+
+
+char sokutei_get_type_of(const int index){
+    return sokutei_counter_definitions[index][31];
+}
+
+
+int sokutei_get_index_of_counter(const char *counter_name){
+    int i;
+    for(i = 0; i < sokutei_number_of_counters; i++){
+        if(sokutei_string_equals(sokutei_counter_definitions[i], counter_name)){
+            return i;
+        }
+    }
+    return SOKUTEI_NOT_FOUND;
+}
+
+
+void sokutei_set_number_of_iterations(const int iterations){
+
+    if(SOKUTEI_MAX_MEASURED_ITERATIONS < iterations){
+        //TODO call error
+        return;
+    }
+
+    number_of_iterations = iterations;
+}
+
+
+int sokutei_add_counter(const char *counter_name, char type){
+
+    if(type != SOKUTEI_INTEGER_TYPE && type != SOKUTEI_FLOAT_TYPE && type != SOKUTEI_INTERVAL_TYPE){
+        return SOKUTEI_NOT_MATCHING_TYPE;
+    }
+
+    if(sokutei_number_of_counters + 1 >= SOKUTEI_MAX_COUNTER_COUNT){
+        return SOKUTEI_COUNTER_LIMIT_REACHED;
+    }
+
+    sokutei_strcpy(sokutei_counter_definitions[sokutei_number_of_counters], counter_name);
+    sokutei_counter_definitions[sokutei_number_of_counters][SOKUTEI_MAX_COUNTER_NAME_LENGTH + 1] = type;
+    sokutei_number_of_counters++;
+}
+
+
+SOKUTEI_INTEGER_COUNTER_TYPE sokutei_integer_get_counter(const char *counter_name){
+    const int index = sokutei_get_index_of_counter(counter_name);
+
+    if(index == SOKUTEI_NOT_FOUND){
+        return SOKUTEI_NOT_FOUND;
+    }
+    if(sokutei_get_type_of(index) != SOKUTEI_INTEGER_TYPE){
+        //TODO call error
+        return SOKUTEI_NOT_MATCHING_TYPE;
+    }
+
+    SOKUTEI_INTEGER_COUNTER_TYPE *counter = sokutei_counters + (index * MAX_SIZE_OF_TYPES);
+
+    return *counter;
+}
+
+
+SOKUTEI_INTEGER_COUNTER_TYPE sokutei_integer_increment_counter(const char *counter_name, const SOKUTEI_INTEGER_COUNTER_TYPE by){
+    const int index = sokutei_get_index_of_counter(counter_name);
+
+    if(index == SOKUTEI_NOT_FOUND){
+        return SOKUTEI_NOT_FOUND;
+    }
+    if(sokutei_get_type_of(index) != SOKUTEI_INTEGER_TYPE){
+        //TODO call error
+        return SOKUTEI_NOT_MATCHING_TYPE;
+    }
+
+    SOKUTEI_INTEGER_COUNTER_TYPE *counter = (void *)sokutei_counters + (index * MAX_SIZE_OF_TYPES);
```

Comment:
Not sure about other compilers, but this breaks on MSVC (Version 19.16.27032.1). Again, what are we trying to achieve here?

We have this `sokutei_counters` variable which points to an array of chars. We want to increment it with some value and then assign it to a pointer of another type. To achieve this, you first cast this variable to `void *` and then attempt to increment it.

This is not going to work, because **pointer arithmetic does not work on void pointers** (https://stackoverflow.com/a/3524270).